### PR TITLE
Add support for HealthCheckNodePort in NodePort BPF

### LINKED ISSUE #8699: Support HealthCheckNodePort in NodePort BPF
See https://github.com/cilium/cilium/blob/5d8a18795370d94d3dfe9ebaf0f5878f0f19b6c6/vendor/k8s.io/api/core/v1/types.go#L3625-L3631 for the description.

Depends on #8698.

### PR DESCRIPTION
**Pull request by gandro**
This adds the `HealthCheckNodePort` field to the service manager package and related types and starts a HTTP server on that port if set. This feature is used by external loadbalancers to determine whether a certain node contains a local backend. When a service has local backends, the HTTP server on the specified port returns `HTTP 200`, otherwise it returns `503 Service Unavailable`. The implementation inside the new `service/healthserver` provides the same JSON output as `kube-proxy`.

The health check server is only enabled for a given service if it has a non-zero `HealthCheckNodePort`. This value may only be set in Kubernetes if a service has `externalTrafficPolicy=Local` and its type is `LoadBalancer`. Our current implementation is a bit more relaxed, it relies on the K8s API to enforce the constraint, since we do not know the original Kubernetes service type.

Review per commit. Ginkgo run-time tests will be added in a separate PR.

Fixes: #8699

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9906)

### CODE REVIEW

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
test-missed-k8s

**Comment by gandro** [comment]
test-missed-k8s

File: pkg/service/healthserver/healthserver.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+// healthServer is a running HTTP health server for a certain service
+type healthServer interface {
+	updateService(*Service)
+	shutdown()
+}
+
+// healthServerFactory creates a new HTTP health server, used for mocking
+type healthServerFactory interface {
```
Comment (brb) [inline pkg/service/healthserver/healthserver.go]: Maybe `s/healthServerFactory/healthHTTPServerFactory/` and `s/newHealthServer/newHealthHTTPServer` to better distinguish from `ServiceHealthServer` (the user of the factory method and the interface)?

File: pkg/service/healthserver/healthserver.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+// ServiceHealthServer manages HTTP health check ports. For each added service,
+// it opens a HTTP server on the specified HealthCheckNodePort and either
+// responds with 200 OK if there are local endpoints for the service, or with
+// 503 Service Unavailable if the service does not have any local endpoints.
+type ServiceHealthServer struct {
+	healthServerByPort  map[uint16]healthServer
+	portRefCount        counter.IntCounter
+	portByServiceID     map[uint16]uint16
```
Comment (brb) [inline pkg/service/healthserver/healthserver.go]: `s/uint16/loadbalancer.ServiceID/`?

**Review by brb — CHANGES_REQUESTED** [review]
LGTM except for a few nits!

Also, would be good to have CI tests for this, and `Documentation/gettingstarted/nodeport.rst` needs update (both can be done in a follow-up).

**Comment by gandro** [comment]
@brb Addressed your nits.

--

[Cilium-PR-Ginkgo-Tests-K8s](https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-K8s/2745/console) failed with `The Service "istio-ingressgateway" is invalid: spec.ports[2].nodePort: Invalid value: 31390: provided port is already allocated`. It looks like the statically defined `nodePort`s in the Istio tests clashed with some randomly allocated ports (I was unable to figure out where the randomly allocated ports were used though).

I while I believe that the failure is *not* caused by this PR (because the port random allocation and rejection both comes from the API server which is not affected by the changes here), it's something we should keep in mind for the future. Adding CI tests which set static `healthCheckNodePorts` will also run into the danger of conflicting with randomly allocated node ports. Not sure if there is something that can be done about it however (except not setting static node ports, and always have them allocated randomly).

**Comment by gandro** [comment]
test-me-please

**Review by brb — APPROVED** [review]

**Comment by gandro** [comment]
test-me-please

**Review by borkmann — APPROVED** [review]

**Comment by gandro** [comment]
test-me-please

**Pull request merged**